### DISCUSSION TITLE
Different BACnet networks support.

### DISCUSSION THREAD

**Discussion by christopherjohnson1535** [JoelBender/bacpypes#513] [Q&A] [answered]
Does BACpypes support communication with devices on different BACnet networks from one BACpypes instance?

**Comment by bbartling** [comment]
Am not the expert here but yeah it should. What is your use case? Are you on an IP LAN network with many different BACnet MSTP subnetworks behind different BACnet routers or something?

**Comment by christopherjohnson1535** [comment]
An on IP LAN and the same physical network as other controllers, just different BACnet networks. I am using a BIPSimpleApplication; is there anything I need to setup to be able to talk to other networks? For example, I want to talk to devices on network 1,2, and 3 from one application

**Reply by bbartling** [reply]
bacpypes3 has an arg for network... maybe legacy bacpypes you can set this somehow in the your bacpypes.ini but not sure

**Reply by christopherjohnson1535** [reply]
Where is this defined in the stack? Is it a part of the application, or apart of the local device object?

**Reply by JoelBender** [reply]
In BACpypes3 it is a configuration parameter of the Network Port Object (Clause 12.56.11) and is used with the Network_Number_Quality property (Clause 12.56.12).  The network number quality is "unknown", "learned", "learned configured", or "configured", and related to the **Network-Number-Is** service (Clause 6.4.20).

**Comment by JoelBender** [comment]
Being on the same physical network isn't quite enough information; you can have multiple IPv4 networks _and_ multiple BACnet networks in the same IPv4 network by using different port numbers (usually 47808, 47809, ...).  In general BACnet devices that are not routers (a.k.a. what I called "simple" and what the standard calls "normal") do not know what BACnet network they are "on" (What-Is-Network-Number and Network-Number-Is are relatively recent additions to the standard).

The `address` portion of the INI file will govern which IPv4 address to use and the port number.  So `192.168.1.15/24:47808` could be used for one application on a network and `192.168.2.12/24:47809` could be used for some other network, provided both of those IPv4 addresses supported by the host (which could be single link or multi-link homing).

If your application is on network 1 and you have routers to networks 2 and 3 then they are all part of the same intranet and everything works as usual.  I'm going to dig around and see what I've got to help running multiple applications (in BACpypes3).

**Reply by christopherjohnson1535** [reply]
All of the controllers are on the same IPV4 network using port 47808. The controllers exists across BACnet networks 1,2, and 3. If I understand you correctly, I would need to spin up different BACnet applications for each network in order to talk to the controllers if they are "simple"? And if they are able to route, then I should be able to talk to them from one application?

**Comment by JoelBender** [comment, marked as answer]
You are correct.  If your application is on network 1 and there is a router from 1 -> 2 and 1 -> 3 (they might be the same device or different devices, and it might be 2 -> 3) then you have an intranet and you only need to be on one of the three at a time.  Some controllers will be "local" (on the same network you are on) and some will be "remote" (one or more hops through routers).  Grab an address on one of the networks, do a global broadcast Who-Is, and see what you get back:
```
$ python3 -m bacpypes3 --address 192.168.1.15/24
>>> whois *:*
...
```
There are cases where there are multiple intranets, groups of BACnet networks that don't talk to each other, but it can be handy to prowl around them using one application.  For that rare case I made [this application](https://github.com/JoelBender/BACpypes3/blob/main/samples/multiple-stacks.py) which might be instructive, or overkill!  It's the curse of seeing lots of strange things in the real world.

**Reply by christopherjohnson1535** [reply]
How is the BACnet network defined? Does it correlate directly with the port, or is explicit? For example, is 47808 always BACnet network 1, 47809 always BACnet network 2?

**Reply by ChristianTremblay** [reply]
It's defined by the people configuring the network. No standard way here.
The standard dictate that all network numbers should be unique. That's it.
By using another port on the same subnet (0xBAC1 for example) you have to use another network number. It is considered as a different network.

**Reply by JoelBender** [reply]
The simplest definition is that a BACnet network is a collection of peer devices that share a common link layer (physical or virtual).  Devices have a unique address within the collection and have the ability to send and receive broadcast messages.  For example, an ARCNET network has addresses like `1`, `2`, and a broadcast address `255`.  For BACnet/IPv4 Virtual Link Layer, devices addresses are like `192.168.1.12:47808` for the BACpypes notation, `("192.168.1.12", 47808)` in Python sockets, with the broadcast address `192.168.1.255` (implicit port number) or `("192.168.1.255", 47808)` in sockets.  Call this network 100.

Consider `192.168.2.0/24` to be another network, a different collection of devices, and call it network 200.  A broadcast on one network cannot reach the other, a.k.a. the two collections have a different broadcast domain.  A router is a network layer device that listens for traffic on one network and forwards it to the other, so as a participant in both networks is has two addresses.  You've made the leap from a network to an intranet.  Clear enough when the distinction is the IPv4 address.

Things get a little muddy when the site designer decides that port 47809 will be used for a different collection of devices that are on the same subnet, so addresses are like `("192.168.1.12", 47809)` and `("192.168.1.255", 47809)`.  Call this network 300.  The BACnet network definition doesn't change, broadcasts on one network will not reach the devices on the other network and you'll need a router to tie the two or more of them together.

Ah, but network numbers are precious!  What if I want more than one IPv4 subnet/port to be the same BACnet network?  Maybe I have one IPv4 subnet per building and all of those subnets should be considered peers?  A BACnet Broadcast Management Device (BBMD) listens for broadcasts on one IPv4/port network and forwards them to a peer on another IPv4/port network which re-broadcasts the PDU.  _This is at the virtual link layer, not the network layer._

My answer run long, a bit [Entish](https://en.wikipedia.org/wiki/Languages_constructed_by_J._R._R._Tolkien#Entish).

**Reply by drewdepriest** [reply]
and assuming a device ID schema that follows <5-digit net number><2-digit device ID> (if you adhere to 99 devices per subnet segment), that leaves roughly **40,000** possible networks on a single UDP port. So, definitely very precious in that any dupe could in theory burn down the house, but also quite manageable if you standardize and enforce a globally unique approach to numbering.

it's been a hot minute since i've asked anyone in networking to open a second UDP port like 47809 and bridge it to 47808, but the last time didn't go so well. 

edit to explain calc: 135 allows for a max 22-bit device ID, which is 2^22 or 4,194,304 possible device IDs on a single UDP port. Using a 5/2 schema can take you up to roughly <41942><99>, so assume ~41,000 possible len=5 network numbers. 

![myprecious](https://github.com/JoelBender/bacpypes/assets/1335002/2a06a529-9b38-4b7c-a395-8f248af77058)

**Comment by bbartling** [comment]
@JoelBender in what applications would someone use the `route_aware` flag? I.E., maybe something like:
`bacpypes.settings.settings['route_aware'] = True`

[bacpypes settings.py sample](https://github.com/JoelBender/bacpypes/blob/master/py34/bacpypes/settings.py)


I do get a bit confused on where a BBMD could be used (to route between different UDP ports or networks?) or using the `route_aware` flag in your app or just make additional bacpypes applications for each unique `network`?

**Reply by JoelBender** [reply]
Being "route aware" just means that the application is all-knowing with respect to the network topology and is not interested in using Who-Is-Router-To-Network to find a path to a router.

Assume a client at `192.168.1.12` is interested in sending a request to a device at address 45 on network 123, or `123:45` in BACpypes.  It could send out a broadcast WIRTN and wait for `192.168.2.13` to reply and cache the result, or it could use its configuration/database and send the PDU to `123:45@192.168.2.13` and bypass that step.  There's no checking to see if that destination is actually a router, or it's a router to that network, or even if the client and the router are part of the same BACnet network!

BBMDs are at the virtual link layer, not the network layer, so they are not routers.

**Reply by JoelBender** [reply]
To answer your question, there is a collection of similar buildings spread across a country and each one has its BACnet IPv4 devices all configured to be network 1, which is fine because they all have different IPv4 addresses, but all of the field bus devices (MS/TP, ARCNET) starting with 2, 3, 4, etc, _...but wait, there's more..._, they all have consecutive device instance numbers starting at 1.  There is no easy way to wrap this up in one BACnet intranet, even with VPN tunnels.

The analysis or asset management application lives in the corporate offices and uses "route aware" keeping all of the networks separate.  It can't use broadcast traffic (Who-Is, Who-Is-Router-To-Network, etc) because it's not "on" those other networks, but unicast is fine, and the servers don't pay attention to the source address of a request (as they shouldn't in most cases).

There is [this thought](https://www.brainyquote.com/quotes/douglas_adams_163199), which can be confirmed with [this code](https://github.com/JoelBender/bmr/wiki).
